### ISSUE TITLE
Can't get $regex to work in find function with feathers-nedb in the background

### ISSUE THREAD

**Issue by Protoss78** [feathersjs/feathers-client#28]
Not sure if this is an error or if I'm using something wrong. I'm trying to find a list of people that contain a certain search string in their names. I wanted to use the $regex operator inside the query for that and extended my already working query in the following way:
```
            var searchString = that.search.split(' ').join('|');
            // Use searchString to build rest of regex
            var regex = new RegExp(searchString, 'i');
            var query = {
              $limit: pageSize,
              $skip: skip,
              $sort: sortObject,
              $or: [{
                name: {
                  $regex: regex
                }
              }, {
                firstName: {
                  $regex: regex
                }
              }, {
                lastName: {
                  $regex: regex
                }
              }]
            };
```
But when I do the find call I get the following error message back:
Uncaught (in promise) $regex operator called with non regular expression

I've checked where the error comes from and found that the exception is thrown in the nedb model.js library. It tries to validate the regex before execution and for some reason the regex that arrives there is an empty object.
Any help is very much appreciated.

**Comment by Protoss78** [comment]
I'm using feathers-socketio as a a transport mechanism by the way. To be honest I'm at the moment not even sure if the issue is filed under the right project.

**Comment by marshallswain** [comment]
Your RegExp object is probably getting parsed to a string (and thus destroyed) in the transport.  You can verify this by creating a simple before-hook in your service that looks something like this:

```
/**
 * Log the query to the console
 */
export default function(){
  return function(hook){
    console.log(hook.params.query);
  };
}
```
That should let you see the query as the database adapter is going to see it.

We recently created an issue for this behavior in the repos for each of our database adapters.  Here's the one for `feathers-nedb`: https://github.com/feathersjs/feathers-nedb/issues/13

**Comment by daffl** [comment]
@marshallswain is probably right. In that hook you can also convert it back to a regex. Maybe something like:

```js
/**
 * Log the query to the console
 */
app.service('todos').before({
  find(hook){
    const query = hook.params.query;
    
    Object.keys.forEach(query, key => {
      const value = query[key];
      if(value && value.$regex) {
        value.$regex = new RegExp(value.$regex);
      }
    });
  }
});
```

**Comment by ekryski** [comment]
@Protoss78 let us know if that works. You should be able monkey patch your regex in a before hook like @marshallswain and @daffl suggested.

We're planning on adding regular expression support via a `$search` param. Progress can be tracked here: https://github.com/feathersjs/feathers-nedb/issues/13 like @marshallswain mentioned. It probably won't happen until Feathers 2.0 is out so a week or two.

If you manage to get it working and feel inclined to make a PR against that repo feel free :D

**Comment by Protoss78** [comment]
Thanks for the quick response and the different code samples. I got it working by creating the RegExp object in a before hook as suggested. I tried the suggested ES6 code but ran into issues when the $regex clause was not included in the query, which ultimately caused problems in my web app. 
Since I'm quite new in the JavaScript world and have no ES6 knowledge so far I ended up with these lines that did the trick for my service:
```
app.use('starters', nedb('starters'));
const starterService = app.service('starters');
starterService.before({
    find(hook) {
        console.log(hook.params.query);
        if (hook.params.query.name) {
            hook.params.query.name.$regex = new RegExp(hook.params.query.name.$regex, 'i');
        }
    }
});
``` 

Since I just started working with feathersjs a few weeks ago in my spare time and still have to finish my transition from the Java world, don't expect any big PR's in the near future. Probably better at the moment. :-D
But I have to give you guys a round of applause for feathersjs. I'm really impressed with the whole framework and its ease of use. :+1: :+1: :+1:
